File: src/iptool.rs
/* iptool-ffi
 
   IP Tool exposed via an FFI for integration into
   libraries for other languages.

   Author: Tim Monfette
   Version: 0.1.0
*/

extern crate iptool_core;
extern crate libc;

use libc::c_char;
use std::ffi::CString;
use std::ffi::CStr;

use iptool_core::ipv4;
use iptool_core::ipv6;

// Validate an IPv4 address
#[no_mangle]
pub extern fn valid_ipv4(s: *const c_char) -> *mut c_char {
    let c_str = unsafe {

        CStr::from_ptr(s)
    };

    let rstr = c_str.to_str().unwrap();
    let ans = ipv4::validate_ipv4(rstr.to_owned());
    let ret = CString::new(ans).unwrap();

    ret.into_raw()
}

// Translate IPv4 into IPv6
#[no_mangle]
pub extern fn ipv4_ipv6(s: *const c_char) -> *mut c_char {
    let c_str = unsafe {

        CStr::from_ptr(s)
    };

    let rstr = c_str.to_str().unwrap();
    let ans = ipv4::ipv4_to_ipv6(rstr.to_owned());
    let ret = CString::new(ans).unwrap();

    ret.into_raw()
}

// Translate IPv4 into Binary
#[no_mangle]
pub extern fn ipv4_bin(s: *const c_char) -> *mut c_char {
    let c_str = unsafe {

        CStr::from_ptr(s)
    };

    let rstr = c_str.to_str().unwrap();
    let ans = ipv4::ipv4_binary(rstr.to_owned());
    let ret = CString::new(ans).unwrap();

    ret.into_raw()
}

// Validate an IPv6 address
#[no_mangle]
pub extern fn valid_ipv6(s: *const c_char) -> *mut c_char {
    let c_str = unsafe {

        CStr::from_ptr(s)
    };

    let rstr = c_str.to_str().unwrap();
    let ans = ipv6::validate_ipv6(rstr.to_owned());
    let ret = CString::new(ans).unwrap();

    ret.into_raw()
}

// Translate IPv6 into IPv4
#[no_mangle]
pub extern fn ipv6_ipv4(s: *const c_char) -> *mut c_char {
    let c_str = unsafe {

        CStr::from_ptr(s)
    };

    let rstr = c_str.to_str().unwrap();
    let ans = ipv6::ipv6_to_ipv4(rstr.to_owned());
    let ret = CString::new(ans).unwrap();

    ret.into_raw()
}

// Translate IPv6 into Binary
#[no_mangle]
pub extern fn ipv6_bin(s: *const c_char) -> *mut c_char {
    let c_str = unsafe {

        CStr::from_ptr(s)
    };

    let rstr = c_str.to_str().unwrap();
    let ans = ipv6::ipv6_binary(rstr.to_owned());
    let ret = CString::new(ans).unwrap();

    ret.into_raw()
}

// Free a string from memory
#[no_mangle]
pub extern fn release_word(s: *mut c_char) {
    unsafe {
        if s.is_null() { return; }
        CString::from_raw(s)
    };
}

File: lib/iptool.js
/* Javascript binding for the IP Tool.
 *
 * Uses the IP Tool FFI to provide all of the
 * functions in a Node.js file.
 *
 * Can replace any of the "console.log" calls to perform
 * any desired action with the result of the FFI function
 * call.
 *
 * Author:
 * Tim Monfette
 */

const ffi = require('ffi');

const lib = ffi.Library('./libiptool', {
  valid_ipv4: ['char *', ['string']],
  ipv4_ipv6: ['char *', ['string']],
  ipv4_bin: ['char *', ['string']],
  valid_ipv6: ['char *', ['string']],
  ipv6_ipv4: ['char *', ['string']],
  ipv6_bin: ['char *', ['string']],
  free_string:['void',['char *']]
});

// Wrapper for valid_ipv4
function valid_ipv4(userInput) {
  const result = lib.valid_ipv4(userInput);
  try {
    return result.readCString();
  } finally {
    lib.free_string(result);
  }
}

// Wrapper for ipv4_ipv6
function ipv4_ipv6(userInput) {
  const result = lib.ipv4_ipv6(userInput);
  try {
    return result.readCString();
  } finally {
    lib.free_string(result);
  }
}

// Wrapper for ipv4_bin
function ipv4_bin(userInput) {
  const result = lib.ipv4_bin(userInput);
  try {
    return result.readCString();
  } finally {
    lib.free_string(result);
  }
}

// Wrapper for valid_ipv6
function valid_ipv6(userInput) {
  const result = lib.valid_ipv6(userInput);
  try {
    return result.readCString();
  } finally {
    lib.free_string(result);
  }
}

// Wrapper for ipv6_ipv4
function ipv6_ipv4(userInput) {
  const result = lib.ipv6_ipv4(userInput);
  try {
    return result.readCString();
  } finally {
    lib.free_string(result);
  }
}

// Wrapper for ipv6_bin
function ipv6_bin(userInput) {
  const result = lib.ipv6_bin(userInput);
  try {
    return result.readCString();
  } finally {
    lib.free_string(result);
  }
}
